runtime/src/utils/buffer.rs

```diff
@@ -111,11 +106,12 @@ impl<B: Blob> Buffer<B> {
         let bytes_to_read = std::cmp::min(self.buffer_size as u64, blob_remaining) as usize;
 
         // Read the data - we only need a single read operation since we know exactly how much data is available
-        self.blob
-            .read_at(&mut self.buffer[..bytes_to_read], self.blob_position)
-            .await?;
+        // Note that the last refill may cause `self.buffer` to have length < `self.buffer_size`
+        // because `bytes_to_read` < `self.buffer_size`.
+        let mut buffer = std::mem::take(&mut self.buffer);
```

Comment:
After this line, I think we need to set `self.buffer_valid_len = 0`. Then I'm g2g.

Reply:
nvm, we set this up top.